### PR TITLE
#0: Feed more data to visualizer

### PR DESCRIPTION
**Pull request by arakhmati**
### Ticket
n/a

### Problem description
Visualizer needs more information from C++

### What's changed
- Commonized how operation and tensor ids are used
- Added device operation id that will be used for tracking device operations
- Removed Operation History

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

#### Pretty Print
```
Capture Start                                     
Function Start: tt::tt_metal::detail::convert_python_tensor_to_tt_tensor
Function End:   tt::tt_metal::detail::convert_python_tensor_to_tt_tensor
Add Tensor: 14                                    
Function Start: ttnn::to_layout                   
    Function Start: Tensor::reshape               
    Function End:   Tensor::reshape               
    Add Tensor: 15                                
    Function Start: Tensor::pad                   
    Function End:   Tensor::pad                   
    Add Tensor: 16                                
    Function Start: Tensor::to                    
    Function End:   Tensor::to                    
    Add Tensor: 17                                
Function End:   ttnn::to_layout                   
Add Tensor: 18                                    
Function Start: Tensor::to                        
    Add Device Buffer                             
    Allocate Device Buffer                        
Function End:   Tensor::to                        
Add Tensor: 19                                    
Function Start: ttnn::add                         
    Function Start: Tensor::to                    
        Add Tensor: 20                            
        Add Device Buffer                         
        Allocate Device Buffer                    
    Function End:   Tensor::to                    
    Add Tensor: 21                                
    Function Start: ttnn::prim::binary            
        Function Start: BinaryDeviceOperation     
            Function Start: tt::tt_metal::create_device_tensor
                Add Device Buffer                 
                Allocate Device Buffer            
            Function End:   tt::tt_metal::create_device_tensor
            Add Tensor: 22                        
            Deallocate All Circular Buffers       
            Add Device Buffer                     
            Allocate Device Buffer                
            Allocate Buffer Allocate              
            Allocate Buffer Allocate              
            Allocate Buffer Allocate              
            Deallocate Device Buffer              
        Function End:   BinaryDeviceOperation     
        Add Tensor: 23                            
    Function End:   ttnn::prim::binary            
    Add Tensor: 24                                
    Deallocate Device Buffer                      
Function End:   ttnn::add                         
Add Tensor: 25                                    
Function Start: Tensor::cpu                       
Function End:   Tensor::cpu                       
Add Tensor: 26                                    
Function Start: Tensor::to                        
Function End:   Tensor::to                        
Add Tensor: 27                                    
Function Start: tt::tt_metal::detail::convert_tt_tensor_to_torch_tensor
Function End:   tt::tt_metal::detail::convert_tt_tensor_to_torch_tensor
Deallocate Device Buffer                          
Capture End  
```

#### Graph
image (3)

### CODE REVIEW

File: ttnn/cpp/ttnn/device_operation.hpp [hunk 1]
```diff
@@ -254,99 +262,188 @@ inline void log_operation(
...
+    };
+
+    if (is_program_cache_enabled) {
+        auto& program = create_or_get_program_from_cache<device_operation_t>(
+            program_cache, program_cache_hit, program_hash, operation_attributes, tensor_args, tensor_return_value);
+
+        enqueue_or_launch_program(program);
+
+        log_device_operation_end<device_operation_t>(
+            operation_id,
```
Comment (mo-tenstorrent) [inline ttnn/cpp/ttnn/device_operation.hpp:407]: Is operation_id unique per op type and primitive_operation_id unique per op instance?

File: ttnn/cpp/ttnn/tensor/tensor.cpp [hunk 2]
```diff
@@ -486,13 +490,14 @@ Tensor Tensor::to(const std::vector<Device*>& workers, const MemoryConfig& mem_c
...
     this->tensor_attributes->update_main_thread_ref_count(workers.at(0), original_tensor_ref_count);
+    device_tensor= tt::tt_metal::set_tensor_id(device_tensor);
     GraphTracker::instance().track_function_end(device_tensor);
     return device_tensor;
 }
 
 Tensor Tensor::cpu(bool blocking, uint8_t cq_id) const {
     ZoneScoped;
-    GraphTracker::instance().track_function_start("Tensor::cpu", blocking);
+    GraphTracker::instance().track_function_start("Tensor::cpu", *this, blocking);
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/tensor/tensor.cpp:500]: good catch with this!

File: ttnn/cpp/ttnn/tensor/tensor.cpp [hunk 3]
```diff
@@ -1158,6 +1177,17 @@ std::vector<Device*> distribute_tensor_to_mesh(const Tensor& tensor, DeviceMesh&
     }
 }
 
+Tensor set_tensor_id(const Tensor& tensor) {
+    if (not GraphTracker::instance().is_enabled()) {
+        return tensor;
+    }
+    std::cout << "Setting tensor id" << std::endl;
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/tensor/tensor.cpp:1184]: log_info instead of std::cout?

**Review by eyonland — APPROVED** [review]

File: ttnn/cpp/ttnn/tensor/tensor.cpp (hunk 3 above)
Comment (arakhmati) [reply ttnn/cpp/ttnn/tensor/tensor.cpp:1219]: Didn't mean to push this

File: ttnn/cpp/ttnn/device_operation.hpp (hunk 1 above)
Comment (arakhmati) [reply ttnn/cpp/ttnn/device_operation.hpp]: @mo-tenstorrent yes. It's used the same way as the id you added in the op_profiler

**Review by tt-asaigal — APPROVED** [review]

**Review by mo-tenstorrent — APPROVED** [review]

File: ttnn/cpp/ttnn/core.hpp [hunk 4]
```diff
@@ -70,4 +70,22 @@ using core::get_memory_config;
 using core::has_storage_type_of;
 using core::pad_to_multiple_of_tile_size;
 using core::set_printoptions;
+
+
+extern std::atomic<std::int64_t> TENSOR_ID;
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/core.hpp:75]: is it possible to put all of it inside of the graph processor?
or you wthink we need it somewhere else?

**Comment by dmakoviichuk-tt** [comment]
Only one question:
1) If you can/want to move tensor_id settings inside of the tracker lets do it.
2) if not, lets create singleton which will hide implementation details instead of extern std::atomic<std::int64_t> TENSOR_ID;
in hpp files.

**Comment by dmakoviichuk-tt** [comment]
And if look at the image we have shared buffer between tensor -1 and 6 and 7. Which might be ok.
6 and 7 might be some local tensors which copy buffers to the outputs.
But I don't understand role of the tensor 5.

**Comment by dmakoviichuk-tt** [comment]
And could you add some instructions how to get your graph and easy print?

File: ttnn/cpp/ttnn/core.hpp (hunk 4 above)
Comment (arakhmati) [reply ttnn/cpp/ttnn/core.hpp:75]: Tensor ID is used in the python code as well. And both C++ and python code need to use the same atomic variable.  Can we leave it as is on this PR and address it later?

**Comment by arakhmati** [comment]
> Only one question:
> 
> 1. If you can/want to move tensor_id settings inside of the tracker lets do it.
> 2. if not, lets create singleton which will hide implementation details instead of extern std::atomicstd::int64_t TENSOR_ID;
>    in hpp files.

As a mentioned in the other comment, I would like this to be addressed in the next PR

**Comment by arakhmati** [comment]
> And could you add some instructions how to get your graph and easy print?

I updated `usage.rst` with an example

File: ttnn/examples/usage/graph_capture.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+with ttnn.manage_device(0) as device:
+    torch_input_tensor = torch.rand((size,), dtype=torch.bfloat16)
+
+    ttnn.graph.begin_graph_capture(ttnn.graph.RunMode.NO_DISPATCH)
+    input_tensor = ttnn.from_torch(torch_input_tensor, layout=ttnn.TILE_LAYOUT, device=device)
+    output_tensor = input_tensor + scalar
+    output_tensor = ttnn.to_torch(output_tensor, torch_rank=1)
+    captured_graph = ttnn.graph.end_graph_capture()
+
+    ttnn.graph.pretty_print(captured_graph)
```
Comment (dmakoviichuk-tt) [inline ttnn/examples/usage/graph_capture.py:26]: nice and easy to use api :)

**Review by dmakoviichuk-tt — APPROVED** [review]

**Pull request merged**